src/analysis/multiSpecies/microbiomeModelingToolbox/additionalAnalysis/analyseObjectiveShadowPrices.m

```diff
@@ -0,0 +1,191 @@
+%                   all nonzero shadow prices should be collected. 
+%                   Allowed inputs: 'Positive','Negative','Nonzero',
+%                   default: 'Nonzero'.
+% numWorkers        Number indicating number of workers in parallel pool
+%                   (default: 0).
+%
+% OUTPUT:
+% ShadowPrices      Table with shadow prices for metabolites that are 
+%                   relevant for each analyzed objective in each analyzed
+%                   model
+% 
+% Author:
+%-Almut Heinken, 07/2018
+
+parser = inputParser();  % Define default input parameters if not specified
+parser.addRequired('modelList', @iscell);
+parser.addRequired('objectiveList', @iscell);
+parser.addParameter('modelIDs',{}, @iscell);
+parser.addParameter('osenseStr','max', @ischar);
+parser.addParameter('SPDef','Nonzero', @ischar);
+parser.addParameter('numWorkers', 0, @(x) isnumeric(x))
+parser.parse(modelList,objectiveList, varargin{:})
+
+modelList = parser.Results.modelList;
+objectiveList = parser.Results.objectiveList;
+modelIDs = parser.Results.modelIDs;
+numWorkers = parser.Results.numWorkers;
+SPDef = parser.Results.SPDef;
+if isempty(modelIDs)
+    for i=1:size(modelList,1)
+        modelIDs{i,1}=strcat('model_',num2str(i));
+    end
+end
+
+% set a solver if not done already
+global CBT_LP_SOLVER
+solver = CBT_LP_SOLVER;
+if isempty(solver)
+    initCobraToolbox;
+    solver = CBT_LP_SOLVER;
+end
+% initialize parallel pool
+if numWorkers > 0
+    % with parallelization
+    poolobj = gcp('nocreate');
+    if isempty(poolobj)
+        parpool(numWorkers)
+    end
+end
+ShadowPrices{1,1}='Metabolite';
+ShadowPrices{1,2}='Objective';
+
+% Compute the solutions for all entered models and objective functions
+solutions={};
+if numWorkers > 0
+    % with parallelization
+    parfor i=1:size(modelList,1)
+        i
+        ShadowPrices{1,i+2}=modelIDs{i,1};
+        model=modelList{i,1};
+        if osenseStr=='max'
+            model.osenseStr='max';
+        elseif osenseStr=='min'
+            model.osenseStr='min';
+        end
+        changeCobraSolver(solver, 'LP');
+        solutionsTemp={};
+        % Compute the solutions for all objectives
+        for j=1:size(objectiveList,1)
+            % optimize for the objective if it is present in the model
+            if ~isempty(find(ismember(model.rxns,objectiveList{j,1})))
+            model=changeObjective(model,objectiveList{j,1});
+            FBAsolution=solveCobraLP(model);
+            solutionsTemp{j,1}=FBAsolution;
+            end
+        end
+        solutions(:,i)=solutionsTemp;
+    end
+else
+    % without parallelization
+    for i=1:size(modelList,1)
+        ShadowPrices{1,i+2}=modelIDs{i,1};
+        model=modelList{i,1};
+        % Compute the solutions for all objectives
+        for j=1:size(objectiveList,1)
+            % optimize for the objective if it is present in the model
+            if ~isempty(find(ismember(model.rxns,objectiveList{j,1})))
```

Comment:
May you use here an internal function?

Reply:
@laurentheirendt I don't understand the comment.